core/drivers/crypto/caam/cipher/caam_cipher_mac.c

```diff
@@ -0,0 +1,883 @@
+		size_todo = fullSize - size_topost;
+	}
+
+	CIPHER_TRACE("FullSize %zu - posted %zu - todo %zu", fullSize,
+		     size_topost, size_todo);
+
+	if (dupdate->src.length) {
+		srcbuf.data = dupdate->src.data;
+		srcbuf.paddr = virt_to_phys(dupdate->src.data);
+
+		if (!srcbuf.paddr) {
+			CIPHER_TRACE("Bad src address");
+			return TEE_ERROR_GENERIC;
+		}
+
+		if (!caam_mem_is_cached_buf(dupdate->src.data,
+					    dupdate->src.length))
+			srcbuf.nocache = 1;
+	}
+
+	if (!size_todo) {
+		/*
+		 * There is no complete block to do:
+		 *   - either input size + already saved data < block size
+		 *   - or no input data and this is the last block
+		 */
+		if (dupdate->last)
+			memcpy(dupdate->dst.data, ctx->ctx.data,
+			       MIN(dupdate->dst.length, ctx->alg->size_ctx));
+
+		goto out;
+	}
+
+	if (dupdate->last) {
+		realloc = caam_set_or_alloc_align_buf(dupdate->dst.data,
+						      &dst_align,
+						      ctx->alg->size_ctx);
+		if (realloc == -1) {
+			CIPHER_TRACE("Dest buffer reallocation error");
+			ret = TEE_ERROR_OUT_OF_MEMORY;
+			goto err;
+		}
+	}
+
+	/* Check if there is some data saved to complete the buffer */
+	if (ctx->blockbuf.filled) {
+		srcbuf.length = ctx->alg->size_block - ctx->blockbuf.filled;
+
+		if (dupdate->last)
+			retstatus = caam_cipher_block(ctx, true, NEED_KEY1,
+						      true, &srcbuf, &dst_align,
+						      CIPHER_BLOCK_IN);
+		else
+			retstatus = caam_cipher_block(ctx, true, NEED_KEY1,
+						      true, &srcbuf, NULL,
+						      CIPHER_BLOCK_IN);
+
+		ctx->blockbuf.filled = 0;
+		if (retstatus != CAAM_NO_ERROR) {
+			ret = TEE_ERROR_GENERIC;
+			goto err;
+		}
+
+		size_todo -= ctx->alg->size_block;
+
+		if (size_todo || size_topost)
+			inc_mac_buffer(&srcbuf, srcbuf.length);
+	}
+
+	srcbuf.length = ctx->alg->size_block;
+
+	while (size_todo) {
+		if (dupdate->last)
+			retstatus = caam_cipher_block(ctx, true, NEED_KEY1,
+						      true, &srcbuf, &dst_align,
+						      CIPHER_BLOCK_NONE);
+		else
+			retstatus = caam_cipher_block(ctx, true, NEED_KEY1,
+						      true, &srcbuf, NULL,
+						      CIPHER_BLOCK_NONE);
+		if (retstatus != CAAM_NO_ERROR) {
+			ret = TEE_ERROR_GENERIC;
+			goto err;
+		}
+
+		size_todo -= ctx->alg->size_block;
+
+		if (size_todo || size_topost)
+			inc_mac_buffer(&srcbuf, srcbuf.length);
+	};
+
+	if (dst_align.data) {
+		if (!dst_align.nocache)
+			cache_operation(TEE_CACHEINVALIDATE, dst_align.data,
+					dst_align.length);
+
+		if (realloc)
+			memcpy(dupdate->dst.data, dst_align.data,
+			       MIN(dupdate->dst.length, dst_align.length));
+	}
+
+out:
+	ret = TEE_SUCCESS;
+
+	if (size_topost) {
+		CIPHER_TRACE("Save input data %zu bytes of %zu", size_topost,
+			     dupdate->src.length);
+
+		struct caambuf cpysrc = {
+			.data = srcbuf.data,
+			.length = size_topost
+		};
+
+		retstatus = caam_cpy_block_src(&ctx->blockbuf, &cpysrc, 0);
+		if (retstatus != CAAM_NO_ERROR)
+			ret = TEE_ERROR_GENERIC;
+	}
+
+err:
+	if (realloc == 1)
```

Comment:
It would have been nice to use an enum or worst case defines instead of these magic numbers.

Reply:
Instead of adding another enum, I made the ```caam_set_or_alloc_align_buf()``` function to return a ```enum caam_status```. The function would also return a boolean through ```bool *realloc``` to indicate if the buffer has been reallocated.